### DISCUSSION TITLE
Idea: add views to compiled binary

### DISCUSSION THREAD

**Discussion by bosix** [confetti-framework/confetti#108] [Ideas]
Hi there :),

first of all: I love the project. Laravel is an amazing framework and the ability to use it in one of my favourite languages is really great.

## "Problem"
One of the advantages of golang is that it compiles into one executable binary with all dependencies in it and so does confetti. The problem with that is that the view templates and other static files are not included in that binary so it's always necessary to ship the binary together with the resources dir (and of course the .env).

## Suggestion
There are golang packages which allow adding static files into the go binary (e.g. https://github.com/gobuffalo/packr). With that, it is possible to add js/css/gohtml/png... files to a so-called "box" before actually compile the binary. All static file data is loaded as a byte array into the go application. 

**Notices**
* The binary will obviously increase by the size of the files so if you add an image with 2mb size the binary size will increase as well.
* With this feature it is not possible to change views/css/... on the fly because any change needs a recompile to take effect. It's recommended to use a file watcher with auto-rebuild and auto-restart (I like https://github.com/codegangsta/gin but it's not that up to date anymore).

## Use Case
When you ship an application to live the fewer files you have to touch or replace the better and easier the deployment is. If you have only one file to switch during an update, your deployment is really fast and more reliable. It's reliable because all dependencies are inside that one file and switching it will automatically update all dependencies. Additionally, If done correctly, your application is faster since requesting static files not longer require disk io.

I have to admit that you can achieve the same effect in terms of reliability if you create a docker image which included all static files already.

**Comment by reindert-vetter** [comment]
Hello Bosix. Thank you for your explanation. At work we deploy the static assets from our own repository. But this only works fine for very large and long-term projects. Until I read your message, I didn't have a solution for how we can make this development-friendly. Especially since many Confetti programmers do not have much experience with Go.

**Reply by bosix** [reply]
Your welcome :)

I think there are 2 options to make this developer friendly.
1. Auto rebuild/compile: there are tools for doing this and devs who are familiar with npm will know the npm file watcher
2. Make it optional: we could disable the feature for development via env or implement a fallback which is always used when not being live. I think packr automatically checks the actual file system if there is no file inside the box.

**Comment by reindert-vetter** [comment]
Why not https://github.com/gobuffalo/packr/tree/master/v2?

![image](https://user-images.githubusercontent.com/16726304/107295517-cdb46d80-6a6f-11eb-9bb0-627c9deeb670.png)

**Reply by bosix** [reply]
The library I mentioned was just an example. Sure, we can use the one you posted as well.

**Comment by reindert-vetter** [comment]
I think it would be nice if there is a file with all the configuration options. And that packr is turned on by default.

**Reply by bosix** [reply]
It should be possible to add config params to the .env if there is any config option necessary. Packr does not need any configuration by default since it automatically fetches everything needed from your go source code directly.

**Reply by reindert-vetter** [reply]
We can also include the view template files by default, but not the rest.

**Comment by reindert-vetter** [comment]
The templates should then be loaded via dependency injection, I think. Then you can also support the current situation.

**Reply by bosix** [reply]
Most generation tools are using go generate for doing pre-compiling things. Packr creates a *.go file in your project which is read like any other go-file. Nothing will change for the developers. The only thing might be that you have to adjust your .gitignore so that the generated go-file is excluded from git.

**Comment by reindert-vetter** [comment]
How are css and javascript files exposed? Would you just send that via the router?

**Reply by bosix** [reply]
Yes, they will be sent via the normal routing. Packr uses boxes as the basic struct for interacting with files. Each box implements `http.FileSystem` so you can expose them like any other files like so:

```golang
box := packr.NewBox("../frontend/dist")
http.Handle("/static", http.FileServer(box))
```

(this is an example out of one of my other projects)

**Reply by reindert-vetter** [reply]
Okay. Besides web.go and api.go we can create an assets.go router file.

**Reply by bosix** [reply]
How are css and js files currently exposed? I couldn't find any clue in the docs and the example have inline css.

**Reply by reindert-vetter** [reply]
There is currently no solution for this from Confetti. But of course you can configure this with nginx or apache.

**Comment by marcelloh** [comment]
Go 1.16 gets embedded files if I'm not mistaken.
See: https://lakefs.io/working-with-embed-in-go/

**Reply by reindert-vetter** [reply]
Thanks for your link. That is indeed a much better solution.

**Comment by bosix** [comment]
@marcelloh looks good. I suggest to wait for go 1.16 then.

(I'm really not up to date with the current changes in go...)

**Reply by marcelloh** [reply]
I've written an embed functionality myself, but with 1.16 I'll change that for a standard

**Comment by reindert-vetter** [comment]
@bosix @marcelloh All templates are now loaded with `//go:embed`. This means that we have to use Go 1.6. 

Pretty cool that you can put everything in 1 file.
_If you don't want to use env files, you just need to export required environment APP_PORT._

**Reply by bosix** [reply]
@reindert-vetter nice :+1: 
I think you mean go 1.16, right :) ?

**Reply by reindert-vetter** [reply]
@bosix You are absolutely right!
